File: 06.py
import time
from libs import *


def isUniqueChars(string: str) -> bool:
    a: dict[str, bool] = {}
    for ascii in string:
        if ascii in a.keys():
            return False
        a[ascii] = True
    return True


def main():
    print("--- Running Day 6 ---")

    start = time.perf_counter()

    try:
        with open("06.txt", "r") as f:
            data = f.readlines()
    except Exception as e:
        raise ValueError("Unable to read file") from e

    input = data[0]
    answer_p1 = 0
    for i in range(0, len(input) - 4):
        marker = "".join(input[i + x] for x in range(4))

        if isUniqueChars(marker):
            answer_p1 = i + 4
            break

    print(f"Output Part 1: {answer_p1}")

    answer_p2 = 0
    for i in range(answer_p1, len(input) - 14):
        marker = "".join(input[i + x] for x in range(14))

        if isUniqueChars(marker):
            answer_p2 = i + 14
            break

    print(f"Output Part 2: {answer_p2}")

    end = time.perf_counter()
    rtime = (end - start) * 1000  # sec -> ms
    print(f"Took {format_runtime(rtime)}\n")


if __name__ == "__main__":
    main()

File: 06.js
const fs = require('fs');
const { performance } = require('perf_hooks');
const { formatRuntime } = require('./libs');

async function readInput() {
  const input = await fs.promises.readFile('06.txt', 'utf-8');
  return input.trim().split('\n');
}

const isUnique = (str) => {
  return new Set(str).size == str.length;
};

const getMarkerStart = (input, startIndex, markerLength) => {
  let answer = 0;
  for (let i = startIndex; i < input.length - markerLength; i++) {
    let marker = '';
    for (let j = 0; j < markerLength; j++) {
      marker += input[i + j];
    }

    if (isUnique(marker)) {
      answer = i + markerLength;
      return answer;
    }
  }
};

async function puzzle() {
  console.log('--- Running Day 6 ---');

  const start = performance.now();

  const input = await readInput();

  let answer = getMarkerStart(input[0], 0, 4);
  console.log('Part 1: ' + answer);

  answer = getMarkerStart(input[0], answer - 1, 14);
  console.log('Part 2: ' + answer);

  const end = performance.now();
  const rtime = end - start;
  console.log('Took:', formatRuntime(rtime));
}

puzzle();
